R/data.table.R

```diff
@@ -1687,12 +1687,12 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
         .ok <- function(q) {
           if (dotN(q)) return(TRUE) # For #5760
           cond = is.call(q) && as.character(q[[1L]]) %chin% gfuns && !is.call(q[[2L]])
-          ans  = cond && (length(q)==2 || identical("na",substring(names(q)[3L],1,2)))
+          ans  = cond && (length(q)==2L || identical("na",substring(names(q)[3L], 1L, 2L)))
           if (identical(ans, TRUE)) return(ans)
-          ans = cond && length(q)==3 && ( as.character(q[[1]]) %chin% c("head", "tail") &&
-                             (identical(q[[3]], 1) || identical(q[[3]], 1L)) ||
-                          as.character(q[[1]]) %chin% "[" && is.numeric(q[[3]]) &&
-                            length(q[[3]])==1 && q[[3]]>0 )
+          ans = cond && length(q) == 3L &&
+            length(q[[3L]]) == 1L && is.numeric(q[[3L]]) && (
+              as.character(q[[1L]]) %chin% c("head", "tail") && q[[3L]] == 1L ||
+                as.character(q[[1L]]) %chin% "["  && q[[3L]] > 0 )
```

Comment:
The logic now seems more natural to me (in particular, I see no reason not to check `is.numeric` and `length` conditions _only_ for the `[` case and not for the `head`/`tail` case -- perhaps a few tests are in order?), and the code reads better too.

Reply:
The use of `identical` may have had `NA` in mind. For the `head / tail` case, `is.numeric` and `length` were being checked via `identical()`. Yes more tests would be great.  Seems like codecov agrees and is not-passing which is good (confirms it's a good exercise).
The line needs a comment saying what it's supposed to be doing, doesn't it!
Can you remove the added spaces around `==` please.  To my eyes there is now lots of separate things all with space between them and it's harder to see the tree of logic.  `=` is already a different character to letters (and is not valid in symbols) so `=` is already a visual separator.  Removing the spaces around `==` allows you to budge the pieces together in this case, so that the remaining spaces are more helpful.
Isn't head and tail with `n>1` now going to pass?  That doesn't seem to be the intent of the original line (only `n==1L` and `n==1` were allowed, iiuc).
An extra set of parens around the left and right of the last `||` might make it clearer?  There was a paren around the left of that `||` in the original but not on its right.